### ISSUE TITLE
Easing

### ISSUE THREAD

**Issue by Martin-Pitt** [gibbok/animatelo#2]
Hi Gibbok, great stuff!

I think your just missing the timing-function from the animate.css

In CSS the property is called 'timing-function', however in WAAPI this is the 'easing' property.

By spec the default in CSS for transitions and animations is "ease".
However in WAAPI the default is "linear" – [AnimationEffectTimingProperties](http://w3c.github.io/web-animations/#dictdef-animationeffecttimingproperties)

You can just pass `{ easing: "ease" }` in the options of your `.animate` call. (This matches `animation-timing-function`/`transition-timing-function` behaviour)
However you can also pass it into the Keyframes as well ('easing' supported in the [Keyframe object](http://w3c.github.io/web-animations/#dictdef-basekeyframe)). Which matches the behaviour of defining the `timing-function` inside the @keyframe animations. (Which animate.css does sometimes, [for example like in the bounce animation](https://github.com/daneden/animate.css/blob/master/animate.css#L39))

**Comment by gibbok** [comment]
Hi @Martin-Pitt , first of all thanks for your positive feedback!

From my understanding you are proposing to add an additional option to animatelo like `{ easing: "ease" }`, which could allow an developer to modify the easing of an animation. Is it correct?

If I understood correctly your question, here my answer:

Unfortunately, this feature won't work for the majority of animation effects included in animatelo, because, similarly to animate.css our animations are composed of multi key-frames.

Example in animatelo:
https://github.com/gibbok/animatelo/blob/master/src/attentionSeekers/bounce.js
Look for property `animationTimingFunction`.

Example in animate.css:
https://github.com/daneden/animate.css/blob/master/source/attention_seekers/bounce.css
Look for property `animation-timing-function`.

This means that every key-frame could have a different value of easing (`animation-timing-function`).

If you are interested in having full control of your animation in animatelo, you could consider creating a custom plugin, which is really straightforward to implement.

Details are provided on this page:
https://github.com/gibbok/animatelo/blob/master/usage.md#extensibility

In case you have a complex CSS animation to be ported to JavaScript, you could use this utility tool I have created:
https://github.com/gibbok/keyframes-tool
The result could be used in a custom plugin in animatelo or directly in WAAPI.

Please let me know if I could provide more information.
Thanks.

**Comment by Martin-Pitt** [comment]
I'm not proposing an additional option for developers.


I mean that the animations appear different to Animate.css because the default easing for the WHOLE animation is missing.


I see that you have actually added per-keyframe `easing`s, but the `easing` for the Animation player _itself_ needs to be accounted for as well.


For example compare [zoomIn](https://github.com/gibbok/animatelo/blob/master/src/zoomingEntrances/zoomIn.js) in animatelo and animate.css, they appear differently, because the initial value for `animation-timing-function` in [CSS spec](https://www.w3.org/TR/css3-animations/#animation-timing-function-property) is `"ease"` but in the [Web Animations spec](https://www.w3.org/TR/web-animations-1/#the-animationeffecttimingproperties-dictionary) it is actually `"linear"`.

I think to fix this you need to set the [defaultOptions](https://github.com/gibbok/animatelo/blob/master/src/animatelo.js#L6) for easing to `"ease"` to match the default behaviour in CSS.

**Comment by gibbok** [comment]
Thanks for your comment and clarification. I hope I understand your question more clearly now.

Comparing zoomIn in animatelo and animate.css I see no differences. Link to source files:
https://github.com/gibbok/animatelo/blob/master/src/zoomingEntrances/zoomIn.js
https://github.com/daneden/animate.css/blob/master/source/zooming_entrances/zoomIn.css

But I have tried to follow your suggestion and adding `easing` for the entire animation.
Unfortunately the result is not satisfactory as the difference from animate.css in easing is very evident with this edit.

IMO currently the easing is applied per key-frame correctly following each single animation from animate.css in animatelo.

If you have a different implementation, please provide a pull request.

Thanks for your time.


Code I have tried:


```js
/*! Animatelo | The MIT License (MIT) | Copyright (c) 2017 GibboK */
; (function (animatelo) {
    'use strict';
    animatelo.version = '1.0.1';

    var _defaultOptions = {
        duration: 1000,
        delay: 0,
        iterations: 1,
        direction: 'normal',
        fill: 'both',
        easing: 'ease'
    },
        _UUID = function () {
            var d = new Date().getTime(),
                uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
                    var r = (d + Math.random() * 16) % 16 | 0;
                    d = Math.floor(d / 16);
                    return (c === 'x' ? r : (r & 0x3 | 0x8)).toString(16);
                });
            return uuid;
        },
        _select = function (selector) {
            var nodeList,
                isNodeList = selector instanceof NodeList,
                isNode = selector instanceof Node,
                isHTMLCollection = selector instanceof HTMLCollection,
                isString = typeof selector === 'string';
            if (isNodeList) {
                nodeList = selector;
            } else if (isNode) {
                nodeList = [selector];
            } else if(isHTMLCollection){
                nodeList = selector;
            } else if (isString) {
                nodeList = document.querySelectorAll(selector)
            } else {
                throw 'selector is invaid';
            }
            return nodeList;
        },
        _validate = function (options) {
            var directionValid = [
                'normal',
                'reverse',
                'alternate',
                'alternate-reverse',
                'initial'
            ],
                fillValid = [
                    'none',
                    'forwards',
                    'backwards',
                    'both',
                    'initial'
                ];
            if (typeof options.duration !== 'number') {
                throw 'parameter duration is invalid';
            }
            if (typeof options.delay !== 'number') {
                throw 'parameter delay is invalid';
            }
            if (typeof options.iterations !== 'number') {
                throw 'parameter iterations is invalid';
            }

            if (typeof options.direction !== 'string' ||
                directionValid.indexOf(options.direction) === -1) {
                throw 'parameter direction is invalid';
            }
            if (typeof options.fill !== 'string' ||
                directionValid.indexOf(options.direction) === -1) {
                throw 'parameter fill is invalid';
            }
        };

    animatelo._animate = function (selector, keyframes, optionsArg) {
        var options = {
            duration: optionsArg && 'duration' in optionsArg ? optionsArg.duration : _defaultOptions.duration,
            delay: optionsArg && 'delay' in optionsArg ? optionsArg.delay : _defaultOptions.delay,
            iterations: optionsArg && 'iterations' in optionsArg ? optionsArg.iterations : _defaultOptions.iterations,
            direction: optionsArg && 'direction' in optionsArg ? optionsArg.direction : _defaultOptions.direction,
            fill: optionsArg && 'fill' in optionsArg ? optionsArg.fill : _defaultOptions.fill,
            easing: optionsArg && 'easing' in optionsArg ? optionsArg.easing : _defaultOptions.easing,
            id: optionsArg && 'id' in optionsArg ? optionsArg.id : _UUID()
        },
            hasUserId = optionsArg && 'id' in optionsArg ? true : false,
            nodeList = _select(selector),
            players = [],
            nodeListArr = [].slice.call(nodeList);
        _validate(options);
        nodeListArr.forEach(function (node, index) {
            var player = node.animate(keyframes, options);
            if (hasUserId) {
                player.id = options.id + '-' + index;
            } else {
                player.id = _UUID();
            }
            players.push(player);
        });
        return players;
    };

})(window.animatelo = window.animatelo || {});

````

**Comment by Martin-Pitt** [comment]
Ok, I finally got some time to put together a test page, a side-by-side comparison.

![image](https://cloud.githubusercontent.com/assets/7586051/25459467/4bab08e4-2ad6-11e7-96d5-b7d8afd730a5.png)


Inspecting things more closely in animation debuggers..

![image](https://cloud.githubusercontent.com/assets/7586051/25459492/60be4c82-2ad6-11e7-8778-e0f3c1bac71f.png)


I notice the bumps are different, so clearly the easing is incorrect, which is why the animation is so wildly different from animate.css.

So I applied a hacky fix...
![image](https://cloud.githubusercontent.com/assets/7586051/25459524/8b7d57d8-2ad6-11e7-82d8-c8a7ed2132bf.png)

Great news.. looks like the animation matches up now:
![image](https://cloud.githubusercontent.com/assets/7586051/25459531/994b9406-2ad6-11e7-9045-4ea406495797.png)


I see where I was wrong now, it's not the default easing of the player that is the issue, but the default easing of the keyframes themselves!

I applied the hack in _animate, looping through keyframes, applying a default easing of "ease" where it wasn't defined.
```
e.forEach(function(kf){if(!kf.easing)kf.easing="ease"});
```

(Also, I noticed a bug in your code, for example at https://github.com/gibbok/animatelo/blob/master/src/attentionSeekers/bounce.js#L7 , `animationTimingFunction` does not exist in the Web Animations spec. It's called something different: `easing`. Otherwise the easing wasn't being recognized when testing in my Chrome browser)

I've uploaded the side-by-side comparison testing page to: http://animatelo-side-by-side.surge.sh/

**Comment by Martin-Pitt** [comment]
1. Proposed code change to main file:
    ```
    _defaultKeyframeOptions = {
        easing: 'ease'
    },
    ```
    ```
    animatelo._animate = function (selector, keyframes, optionsArg) {
        …
        keyframes.forEach(function (keyframe) {
            keyframe.easing = 'easing' in keyframe? keyframe.easing : _defaultKeyframeOptions.easing;
        });
        …
    ```
2. Change all instances of `animationTimingFunction:` in the effects to `easing:`

**Comment by gibbok** [comment]
Solved at release 1.0.2
https://github.com/gibbok/animatelo/releases/tag/1.0.3
